### PR TITLE
Complete BES OTA after reboot reconnect

### PR DESCRIPTION
**Pull request by PhilippeFerreiraDeSousa**
## Summary

- restore the BES wire contract on both iOS and Android: `success` becomes terminal `step_complete`, while `update` remains nonterminal
- cap nonterminal BES step progress at 95%, including raw 98-100%, because whole-image CRC and apply happen afterward; weighted multi-step overall progress remains below 100%
- arm build 37+ reboot recovery only from terminal status, then complete after the BES-specific disconnect/reconnect edge
- prevent an earlier APK process reconnect from satisfying the later BES reboot lifecycle
- retain the former `progress-legacy.tsx` lost-`FINISHED` disconnect fallback only for the sticky pre-37 protocol profile
- remove the unreliable post-reboot BES-version request/comparison loop and its unreleased verification-failure copy

## Root cause

The BES protocol already has an exact terminal signal. `sr_adota.B.type:"update"` is transfer progress; after whole-image CRC and apply validation, the happy path emits `sr_adota.B.type:"success"`, then schedules the reboot. An error path emits `error` instead.

Staging parsed the explicit `success` message but converted it to `ota_status.status:"in_progress"` at 100% on both platforms. The coordinator then attempted to recover by treating numeric 100% as reboot-ready. That is unsafe because the native bridges round raw 98-99% to 100%, and even raw update 100% precedes final CRC/apply.

The corrected contract is:

- `update` -> `in_progress`, BES step progress capped at 95%; weighted overall progress remains below 100%
- `success` -> `step_complete`, 100%
- `error` / `fail` -> `failed`
- terminal status -> expected disconnect -> reconnect -> complete

No percentage threshold determines success.

## Protocol and device evidence

The BES source confirms the ordering:

- `apps/xysmart/m8_uart_ota.cpp` calculates ordinary update progress as `received * 100 / total_size`
- `services/ota/bes_ota/src/ota_control.cpp` emits update progress during segment verification, then performs whole-image CRC, and only the valid image-apply branch emits `OTA_STATE_SUCCESS`
- `apps/xysmart/m8_cmd.cpp` serializes `progress` only for `type:"update"`; `success` deliberately has no progress field

The captured iPhone log demonstrates why numeric 100% is not terminal and that the later explicit success was being erased:

```text
Aug  5 14:39:47.862623 Mentra(Mentra.debug.dylib)[17625] <Info>: BLE_TRACE direction=phone_to_app layer=sdk_event_dispatch source=sendTypedMessage(_:body:)(MentraBluetoothSDK/Bridge.swift:561) type=ota_status payload={"type":"ota_status","phase":"install","step_percent":100,"session_id":"e257e639","current_step":2,"total_steps":2,"overall_percent":100,"step_type":"bes","status":"in_progress"}
Aug  5 14:39:47.863926 Mentra(React)[17625] <Info>: 'CORE:', 'LIVE: 📱 BES OTA progress via sr_adota - type: update, raw: 98%, rounded: 100%'
Aug  5 14:39:49.778987 Mentra(Mentra.debug.dylib)[17625] <Info>: BLE_TRACE direction=phone_to_app layer=sdk_event_dispatch source=sendTypedMessage(_:body:)(MentraBluetoothSDK/Bridge.swift:561) type=ota_status payload={"step_type":"bes","type":"ota_status","overall_percent":100,"session_id":"e257e639","current_step":2,"phase":"install","total_steps":2,"step_percent":100,"status":"in_progress"}
Aug  5 14:39:49.780230 Mentra(React)[17625] <Info>: 'CORE:', 'LIVE: 📱 BES OTA progress via sr_adota - type: success, raw: 0%, rounded: 0%'
```

The later failed validation also showed that the hardware update and reboot succeeded while version metadata remained empty:

```text
08-06 17:58:42.719  3139  3194 I BesOtaManager: BES firmware update SUCCESS! BES will reboot.
Aug  6 10:58:44.707863 Mentra(React)[20695] <Info>: [OTA_PROGRESS] BES restart disconnect observed — awaiting reconnect verification
Aug  6 10:59:49.436825 Mentra(React)[20695] <Info>: [OTA_PROGRESS] connect-edge: false->true, flipping sawReconnectEdge=true
Aug  6 10:59:49.436844 Mentra(React)[20695] <Info>: [OTA_PROGRESS] connect-edge: BES reboot reconnect observed — no ota_query_status
Aug  6 10:59:52.832444 Mentra[20695] <Info>: BLE_TRACE direction=glasses_to_phone layer=sdk_ble_event source=processJsonObject(_:)(MentraBluetoothSDK/MentraLive.swift:2416) type=version_info_3 payload={"bt_mac_address":"","type":"version_info_3","mtk_fw_version":"MentraLive_20260709","bes_fw_version":""}
Aug  6 11:00:49.531372 Mentra(React)[20695] <Info>: [OTA_PROGRESS] BES version still unverified after 12 attempts, failing
```

Postflight ADB showed ASG `39.0` alive on active slot `_b`, and the glasses remained operational.

## Compatibility

- Build 37+ consumes native `step_complete` directly and never treats numeric progress as terminal.
- Pre-37 receives the corresponding legacy `FINISHED` mapping.
- The old legacy screen also accepted BES install -> disconnect -> reconnect when `FINISHED` was lost. The merged coordinator preserves that fallback only when the protocol profile was selected as legacy at flow start.

## Validation

- `bun run test --runInBand --silent src/__tests__/otaInstallCoordinator.test.ts src/app/ota/__tests__/deriveOtaDisplayState.test.ts` - 106/106 passed
- `bun run compile` - passed
- `./scripts/check-android-compile.sh bluetooth-sdk` - `BUILD SUCCESSFUL`
- `:mentra-bluetooth-sdk:testDebugUnitTest --tests com.mentra.bluetoothsdk.sgcs.BesOtaProgressMapperTest` - `BUILD SUCCESSFUL`
- `swiftc -parse mobile/modules/bluetooth-sdk/ios/Source/sgcs/MentraLive.swift` - passed
- `swiftformat --lint mobile/modules/bluetooth-sdk/ios/Source/sgcs/MentraLive.swift` - passed
- Prettier check and `git diff --check` - passed
- focused ESLint passes for the changed coordinator; the existing coordinator test file still reports its six pre-existing restricted-relative-import findings

Regression coverage includes explicit success, raw update 100%, rounded raw 98% -> BES step display 95%, modern disconnect/reconnect rejection without terminal status, APK reconnect isolation, pre-37 lost-`FINISHED` recovery, lockout behavior, retry reset, and empty/stale BES metadata.

---

> [!NOTE]
> **Medium Risk**
> Touches OTA state machine and native BLE mapping on both platforms; behavior change is intentional but incorrect edge handling could still strand or prematurely complete BES updates.
> 
> **Overview**
> Fixes BES firmware OTA completion by aligning native `sr_adota` handling and the install coordinator with the wire protocol instead of treating transfer progress as success.
> 
> **Native (Android/iOS):** `type:"success"` now maps to terminal **`step_complete`** (was `in_progress` at 100%). Non-terminal **`update`** progress is capped at **95%** display (including raw/rounded 98–100%) because CRC/apply still run afterward. Shared **`mapBesOtaProgress`** on Android centralizes rounding and terminal vs progress rules.
> 
> **Coordinator / UI state:** Reboot recovery arms only from **`step_complete` / `complete`** (not percent ≥ 100 or stale legacy `FINISHED` in modern sessions). Completion requires the **BES-specific disconnect → reconnect** path; generic **`sawReconnectEdge`** no longer finishes BES early (e.g. after APK ASG restart). **Post-reboot BES version polling/comparison** and its failure copy are removed—success is the explicit success signal plus reboot edge. **Pre-37** keeps the lost-`FINISHED` install disconnect fallback only for the sticky legacy protocol profile.
> 
> Tests updated for the new contract and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 64ff235494c614acfb999cda65ae71f423190338. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: mobile/modules/engine/src/services/OtaInstallCoordinator.ts [hunk 1]
```diff
@@ -1108,70 +1093,16 @@ class OtaInstallCoordinator {
...
-      if (compareVersions(current, this.besTargetVersion) !== 0) return
-      console.log(
-        `[OTA_PROGRESS] BES restart verified by version_info (${current}) — completing despite stale OTA progress`,
-      )
-    } else if (legacySession) {
-      console.log("[OTA_PROGRESS] legacy BES restart reconnected without target metadata — completing by reboot edge")
-    } else {
-      return
-    }
+    console.log("[OTA_PROGRESS] BES restart reconnected — completing by reboot edge")
```
Comment (aisraelov) [inline mobile/modules/engine/src/services/OtaInstallCoordinator.ts:1109]: **P2 — Don’t complete from rounded 100% progress**

`isBesRestartExpected()` also latches when `stepPercent >= 100` / `overallPercent >= 100`, but both Mentra Live bridges round `sr_adota` to the nearest 5%. A raw 98–99% `update` therefore reaches the coordinator as `in_progress` with `stepPercent: 100`. With version verification removed, an unrelated BLE drop at that point followed by reconnect reaches this method and reports success even though BES never sent raw 100/`success` (apply acceptance). Please preserve an unambiguous terminal/reboot-ready signal—for example, cap rounded nonterminal progress at 95 or emit a distinct status/flag only for raw 100/`success`—and add a 98% → disconnect → reconnect regression test.

**Review by aisraelov — COMMENTED** [review]
One correctness issue remains before I can approve.

**Review by aisraelov — APPROVED** [review]
Re-reviewed the updated BES OTA wire contract and reconnect gating. The explicit terminal success mapping, 95% cap for nonterminal progress, and BES-specific disconnect/reconnect requirement address the earlier false-completion path. Focused coordinator/display tests, the Android mapper test, and compile checks pass.

File: mobile/modules/engine/src/services/OtaInstallCoordinator.ts [hunk 2]
```diff
@@ -1093,85 +1076,40 @@ class OtaInstallCoordinator {
...
-        otaStatus.status === "complete" ||
-        otaStatus.stepPercent >= 100 ||
-        otaStatus.overallPercent >= 100)
+      (otaStatus.status === "step_complete" || otaStatus.status === "complete")
     const progressReadyToRestart =
-      otaProgress?.currentUpdate === "bes" &&
-      otaProgress.stage === "install" &&
-      (otaProgress.status === "FINISHED" || otaProgress.progress >= 100)
-    return statusReadyToRestart || progressReadyToRestart
+      otaProgress?.currentUpdate === "bes" && otaProgress.stage === "install" && otaProgress.status === "FINISHED"
```
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply mobile/modules/engine/src/services/OtaInstallCoordinator.ts]: Addressed in 64ff235494. The otaProgress FINISHED compatibility source is now gated by the sticky legacySession profile; build 37+ recovery must be authorized by terminal otaStatus. Added a regression with modern nonterminal ota_status plus stale legacy FINISHED, followed by disconnect/reconnect; it cannot complete and queries status instead.

File: mobile/modules/bluetooth-sdk/ios/Source/sgcs/MentraLive.swift [hunk 3]
```diff
@@ -3013,34 +3013,33 @@ class MentraLive: NSObject, SGCManager {
...
                     besOtaErrorMessage = bodyObj["message"] as? String ?? "BES update failed"
-                    lastBesOtaProgress = -1 // Reset for next OTA
                 } else {
-                    // BES success/100 proves transfer/apply acceptance, not which image booted.
-                    // ASG sends the existing terminal ota_status only after a fresh sr_syvr
-                    // exactly matches the target version after reboot.
-                    besOtaProgressVal = (type == "success" || rawProgress >= 100) ? 100 : progress
-                    lastBesOtaProgress = besOtaProgressVal
+                    // A raw update:100 precedes whole-image CRC/apply and is not terminal.
+                    besOtaProgressVal = min(progress, 95)
```
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply mobile/modules/bluetooth-sdk/ios/Source/sgcs/MentraLive.swift:3029]: The 96-97% observation is correct, but this is intentional weighted overall progress, not a terminal signal: APK+MTK+BES at BES step 95 computes 50 + floor(95*50/100) = 97; APK+BES computes 96. Both Android and iOS remain strictly below 100, and completion depends only on terminal status. Capping overall at 95 would make multi-step progress inaccurate. The PR description now distinguishes BES step progress (capped at 95) from weighted overall progress (below 100).

File: mobile/modules/engine/src/services/OtaInstallCoordinator.ts [hunk 4]
```diff
@@ -1150,6 +1158,16 @@ class OtaInstallCoordinator {
       return
     }
 
+    const currentVersion = useGlassesStore.getState().besFirmwareVersion.trim()
+    if (this.besVersionResponseObserved && !currentVersion) {
```
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply mobile/modules/engine/src/services/OtaInstallCoordinator.ts]: Obsolete on the current head: the post-reboot version_info polling/grace-period implementation was removed. BES completion now uses explicit protocol success plus the BES-specific disconnect/reconnect edge.

File: mobile/modules/engine/src/services/OtaInstallCoordinator.ts [hunk 5]
```diff
@@ -1108,70 +1093,16 @@ class OtaInstallCoordinator {
...
 
-  /**
-   * Complete a phone-observed BES reboot when fresh version_info matches the
-   * manifest target. Frozen legacy manifests did not always expose that target
-   * through the mobile read model, so truly legacy flows retain the old route's
-   * disconnect/reconnect completion fallback.
-   */
-  private reconcileBesRestart(currentVersion: string, legacySession: boolean): void {
+  /** Complete from the same install -> disconnect -> reconnect invariant as the legacy screen. */
+  private completeBesRestartRecovery(): void {
```
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply mobile/modules/engine/src/services/OtaInstallCoordinator.ts:1107]: Not retained on the current design. BES type=success is emitted only after whole-image CRC and valid apply, then reboot is scheduled; captured hardware logs show successful update/reboot while version_info_3 still reports an empty BES version. Version metadata is therefore not a reliable fail-closed gate. Current head requires explicit terminal success plus the BES-specific disconnect/reconnect edge.

File: mobile/modules/engine/src/services/otaInstallPolicy.ts [hunk 6]
```diff
@@ -127,5 +123,4 @@ export const OtaProgressMessages = {
   stalledOrStuck: "Update may have failed. Ensure glasses have internet access and try again.",
   globalTimeout: "Update took too long. Please try again.",
   sendOtaStartFailed: "Failed to communicate with glasses.",
-  besVersionVerificationFailed: "Glasses restarted, but the updated firmware could not be verified. Please try again.",
```
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply mobile/modules/engine/src/services/otaInstallPolicy.ts:130]: No released engine compatibility is being removed: besVersionVerificationFailed was introduced only in the unreleased intermediate implementation on this PR and is removed with that implementation. It was not present in a published engine package.

File: mobile/src/__tests__/otaInstallCoordinator.test.ts [hunk 7]
```diff
@@ -705,85 +689,53 @@ describe("OtaInstallCoordinator BES reboot recovery", () => {
     expect(bluetoothSdkMock.requestVersionInfo).not.toHaveBeenCalled()
   })
 
-  it("holds stale BES 100% and Continue until a semantically equivalent target version arrives", async () => {
+  it("completes immediately on the reboot edge without gating on BES version metadata", async () => {
```
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply mobile/src/__tests__/otaInstallCoordinator.test.ts:730]: The PR description and validation section have been updated to match the current reboot-edge implementation and current 106/106 focused test count. Removed version-polling behavior is no longer claimed.

File: mobile/modules/engine/src/services/OtaInstallCoordinator.ts (hunk 1 above)
Comment (PhilippeFerreiraDeSousa) [reply mobile/modules/engine/src/services/OtaInstallCoordinator.ts:1109]: Addressed on the current head. Both native bridges keep update messages nonterminal, cap BES step display at 95 even for raw 98-100, and map only explicit success to step_complete. Coordinator regression coverage confirms nonterminal 100 cannot authorize disconnect/reconnect completion.
